### PR TITLE
rustc: Remove `dylib` crate type from most rustc crates

### PR DESCRIPTION
**Pull request by Zoxc**
Revival of https://github.com/rust-lang/rust/pull/56987

cc @alexcrichton 
r? @michaelwoerister

### CODE REVIEW

**Comment by alexcrichton** [comment]
r=me on this, sorry I was never able to get to the bottom of the issues!

**Comment by michaelwoerister** [comment]
Thanks for looking into this, @Zoxc!
r? @alexcrichton

File: src/librustc_codegen_llvm/lib.rs [hunk 1]
```diff
@@ -39,13 +39,14 @@ extern crate rustc_incremental;
 extern crate rustc_codegen_utils;
 extern crate rustc_codegen_ssa;
 extern crate rustc_fs_util;
+#[allow(unused_extern_crates)]
+extern crate rustc_driver;
```
Comment (petrochenkov) [inline src/librustc_codegen_llvm/lib.rs]: `unused_extern_crates` can also be silenced with `extern crate rustc_driver as _;`

**Comment by Zoxc** [comment]
Blocked on https://github.com/rust-lang/rust/pull/59752 landing in beta.

**Comment by Zoxc** [comment]
This is now unblocked.

File: src/libfmt_macros/Cargo.toml [hunk 2]
```diff
@@ -11,3 +11,4 @@ crate-type = ["dylib"]
 
 [dependencies]
 syntax_pos = { path = "../libsyntax_pos" }
+
```
Comment (bjorn3) [inline src/libfmt_macros/Cargo.toml]: Why this extra line?

**Comment by Zoxc** [comment]
This is still blocked https://github.com/rust-lang/rust/pull/59752, since we need it in the bootstrap compiler.

**Comment by cup** [comment]
https://github.com/rust-lang/rust/issues/61978

**Comment by cup** [comment]
https://github.com/rust-lang/rust/issues/61978
https://github.com/rust-lang/rust/pull/56987
https://github.com/rust-lang/rust/pull/59752

https://github.com/rust-lang/rust/pull/59800

**Comment by eddyb** [comment]
*(originally posted as https://github.com/rust-lang/rust/pull/56987#issuecomment-498889066)*

@alexcrichton Has shipping `.rmeta` instead of `.rlib` been considered?
We would have to either include MIR for everything and codegen at the end (potentially subopyimal) or just throw away the `.rlib` and somehow require the final `.so` crate to be linked in order to use anything from the component crates.
For the latter, it could be as simple as a check like "at least one of the dependency crates must claim to contain each symbol referenced *and* be linkable".

**Comment by alexcrichton** [comment]
@eddyb I don't really think it's been considered per se but so long as the number of dynamic libraries we ship goes down and everything largely continues working as is I don't personally mind really how it all happens. This was just one way to achieve the goal but there's surely others!

**Comment by eddyb** [comment]
@alexcrichton Right, I was replying to concerns (that I heard of elsewhere) along the lines of there being duplication between the object files in `.rlib`s and the final `.so`, resulting in increased download sizes.

Whereas with `.rmeta` (*if* we can pull that off), the libraries would still be usable as far as Rust compilation is concerned, but only the `.so` would contain the compiled machine code.

**Comment by Zoxc** [comment]
This is now unblocked.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/62443#issuecomment-508931683, @bors r- rollup=never

**Comment by Zoxc** [comment]
I guess plugins have to link to `rustc_driver` now to avoid multiple copies of rustc libraries.

**Comment by mati865** [comment]
Final perf results: https://perf.rust-lang.org/compare.html?start=dbeed58adee2ef046b46b252980f86672f9bfc4c&end=dd2e8040a35883574ae0c4cc7a4e887ecb66469c&stat=instructions%3Au

**Pull request merged**